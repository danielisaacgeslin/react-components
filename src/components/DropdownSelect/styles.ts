import { StyleSheet } from 'aphrodite/no-important';
import { Color } from 'utils/dist/models/general';
import { hexToRgba } from 'utils/dist/utils';

export default StyleSheet.create({
  container: {
    display: 'inline-block',
    fontSize: 12
  },
  dropdownContainer: {
    position: 'relative'
  },
  dropdownBody: {
    position: 'absolute',
    left: 0,
    top: 5,
    minWidth: '100%',
    boxSizing: 'border-box',
    background: Color.WHITE,
    padding: 12,
    boxShadow: '0 1px 5px 0 ' + hexToRgba(Color.GRAY, 25),
    zIndex: 300
  },
  box: {
    border: '1px solid ' + hexToRgba(Color.GRAY, 25),
    borderRadius: 5
  },
  clear: {
    appearance: 'none',
    padding: '10px 0 0',
    border: 0,
    fontSize: 14,
    fontWeigh: 300,
    background: 'transparent',
    color: Color.GRAY,
    cursor: 'pointer'
  },
  option: {
    display: 'flex',
    alignItems: 'center',
    cursor: 'pointer',
    marginBottom: 10
  },
  optionLabel: {
    display: 'inline-block',
    marginLeft: 7,
    textTransform: 'initial',
    whiteSpace: 'nowrap'
  },
  chip: {
    cursor: 'pointer',
    paddingLeft: 7
  },
  button: {
    appearance: 'none',
    padding: '0 7px',
    lineHeight: '32px',
    color: Color.GRAY,
    backgroundColor: Color.WHITE
  },
  buttonWrapper: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between'
  },
  caret: {
    marginLeft: 10
  },
  caretRotate: {
    transform: 'rotate(180deg)'
  }
});
