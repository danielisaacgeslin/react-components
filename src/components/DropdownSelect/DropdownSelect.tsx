import * as React from 'react';
import { css } from 'aphrodite/no-important';

import ClickOutsideListener from '../ClickOutsideListener';
import Icon from '../Icon';
import Checkbox from '../Checkbox';
import styles from './styles';

export interface IDropdownSelectProps {
  label: string;
  initialValue: string[];
  optionList: { value: string; label: string }[];
  showChips?: boolean;
  onClose?: (value: string[]) => void;
  onOpen?: () => void;
}

export interface IDropdownSelectState {
  value: string[];
  tempValue: string[];
  isOpen: boolean;
}

export default class DropdownSelect extends React.PureComponent<IDropdownSelectProps, IDropdownSelectState> {
  public static defaultProps = {
    label: 'select options',
    initialValue: [],
    optionList: [],
    showChips: true,
    onOpen: /* istanbul ignore next */ () => null,
    onClose: /* istanbul ignore next */ () => null
  };

  public state: IDropdownSelectState = {
    value: [],
    tempValue: [],
    isOpen: false
  };

  public ref = React.createRef<HTMLDivElement>();

  public componentDidMount() {
    const { optionList, initialValue, onClose } = this.props;
    const value = initialValue.filter(v => !!optionList.find(option => option.value === v));
    this.setState({ value, tempValue: value });
    onClose(value);
  }

  public toggleOpen = (open: boolean) => () => {
    const { onClose, onOpen } = this.props;
    const { tempValue, isOpen } = this.state;
    if (open === isOpen) return;
    this.setState({ isOpen: open });
    if (!open) {
      this.setState({ value: tempValue });
      onClose(tempValue);
    } else {
      onOpen();
    }
  };

  public toggleOption = (selected: string, isTemp: boolean) => () => {
    const { onClose } = this.props;
    const { tempValue, value } = this.state;
    const newValue = tempValue.includes(selected) ? tempValue.filter(v => v !== selected) : [...tempValue, selected];
    this.setState({ tempValue: newValue, value: isTemp ? value : newValue });
    if (!isTemp) onClose(newValue);
  };

  public clear = () => {
    this.setState({ tempValue: [] });
  };

  public render() {
    const { label, optionList, showChips } = this.props;
    const { isOpen, value, tempValue } = this.state;
    const optionMap = optionList.reduce((total, curr) => ({ ...total, [curr.value]: curr.label }), {});
    return (
      <div key="DropdownSelect" className={css(styles.container)}>
        <ClickOutsideListener onClickOutside={this.toggleOpen(false)}>
          <button type="button" className={css(styles.button, styles.box)} onClick={this.toggleOpen(true)}>
            <span className={css(styles.buttonWrapper)}>
              {(showChips || !value.length) && label}
              {!showChips && !!value.length && `${label} (${value.length})`}
              <Icon icon="chevron-down" size="sm" className={css(styles.caret, isOpen && styles.caretRotate)} />
            </span>
          </button>
          {showChips &&
            value.map(itemValue => (
              <span
                key={itemValue}
                onClick={
                  !isOpen
                    ? this.toggleOption(itemValue, false)
                    : () => {
                        /** */
                      }
                }
                className={css(styles.chip)}
                data-testid="chip"
              >
                {optionMap[itemValue]} <Icon icon="times" />
              </span>
            ))}
          <div className={css(styles.dropdownContainer)}>
            {isOpen && (
              <div className={css(styles.dropdownBody, styles.box)}>
                {optionList.map(option => (
                  <div className={css(styles.option)} key={option.value} onClick={this.toggleOption(option.value, true)} data-testid="dropdown-option">
                    <Checkbox checked={tempValue.includes(option.value)} />
                    <span className={css(styles.optionLabel)}>{String(option.value)}</span>
                  </div>
                ))}
                <button className={css(styles.clear)} type="button" onClick={this.clear}>
                  Clear
                </button>
              </div>
            )}
          </div>
        </ClickOutsideListener>
      </div>
    );
  }
}
