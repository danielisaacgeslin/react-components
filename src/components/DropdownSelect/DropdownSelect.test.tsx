import * as React from 'react';
import { shallow } from 'enzyme';
import { create } from 'react-test-renderer';

import DropdownSelect, { IDropdownSelectProps } from './DropdownSelect';

describe('CaseDetail Component', () => {
  let props: IDropdownSelectProps;

  beforeEach(() => {
    global.console.error = () => {
      /** */
    };
    props = {
      label: 'dropdown',
      optionList: [{ label: 'one', value: '1' }, { label: 'two', value: '2' }],
      initialValue: [],
      onOpen: jest.fn(),
      onClose: jest.fn()
    };
  });

  it('should render closed', () => {
    const Component = shallow<DropdownSelect>(<DropdownSelect {...props} />);
    expect(create(Component.getElement()).toJSON()).toMatchSnapshot();
  });

  it('should render with initial value', () => {
    props.initialValue = [props.optionList[0].value];
    const Component = shallow<DropdownSelect>(<DropdownSelect {...props} />);
    Component.find('button')
      .first()
      .simulate('click');
    expect(create(Component.getElement()).toJSON()).toMatchSnapshot();
  });

  it('should render without showing chips', () => {
    props.initialValue = [props.optionList[0].value];
    props.showChips = false;
    const Component = shallow<DropdownSelect>(<DropdownSelect {...props} />);
    expect(create(Component.getElement()).toJSON()).toMatchSnapshot();
  });

  it('should clear', () => {
    const Component = shallow<DropdownSelect>(<DropdownSelect {...props} />);
    Component.find('button')
      .first()
      .simulate('click');
    Component.find('[data-testid="dropdown-option"]')
      .at(1)
      .simulate('click');
    Component.find('button')
      .last()
      .simulate('click');
    Component.instance().toggleOpen(false)();
    expect(props.onClose).lastCalledWith([]);
  });

  it('should toggle option', () => {
    const Component = shallow<DropdownSelect>(<DropdownSelect {...props} />);
    Component.find('button')
      .first()
      .simulate('click');
    Component.find('[data-testid="dropdown-option"]')
      .at(1)
      .simulate('click');
    Component.find('[data-testid="dropdown-option"]')
      .at(1)
      .simulate('click');
    Component.instance().toggleOpen(false)();
    expect(props.onClose).lastCalledWith([]);
  });

  it('should trigger onOpen if it was closed', () => {
    const Component = shallow<DropdownSelect>(<DropdownSelect {...props} />);
    Component.find('button')
      .first()
      .simulate('click');
    Component.find('button')
      .first()
      .simulate('click');
    expect(props.onOpen).toBeCalledTimes(1);
    expect(props.onClose).toBeCalledTimes(1);
  });

  it('should trigger onClose', () => {
    const Component = shallow<DropdownSelect>(<DropdownSelect {...props} />);
    Component.find('button')
      .first()
      .simulate('click');
    Component.find('[data-testid="dropdown-option"]')
      .at(1)
      .simulate('click');
    Component.instance().toggleOpen(false)();
    expect(props.onClose).toBeCalledWith([props.optionList[1].value]);
  });

  it('should trigger onClose pressing on a chip', () => {
    const Component = shallow<DropdownSelect>(<DropdownSelect {...props} />);
    Component.find('button')
      .first()
      .simulate('click');
    Component.find('[data-testid="dropdown-option"]')
      .at(1)
      .simulate('click');
    Component.instance().toggleOpen(false)();
    Component.find('[data-testid="chip"]')
      .first()
      .simulate('click');
    expect(props.onClose).toBeCalledWith([]);
    expect(props.onClose).toBeCalledTimes(3);
  });

  it('should trigger NOT onClose pressing on a chip if dropdown is open', () => {
    const Component = shallow<DropdownSelect>(<DropdownSelect {...props} />);
    Component.find('button')
      .first()
      .simulate('click');
    Component.find('[data-testid="dropdown-option"]')
      .at(1)
      .simulate('click');
    Component.instance().toggleOpen(false)();
    Component.find('button')
      .first()
      .simulate('click');
    Component.find('[data-testid="chip"]')
      .first()
      .simulate('click');
    expect(props.onClose).toBeCalledTimes(2);
  });
});
