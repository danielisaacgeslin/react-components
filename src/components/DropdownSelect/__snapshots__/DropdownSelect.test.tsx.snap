// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaseDetail Component should render closed 1`] = `
.container {
  display: inline-block;
  font-size: 12px;
}

.button-box {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 0 7px;
  line-height: 32px;
  color: #D5DEF1;
  background-color: #FFF;
  border: 1px solid rgba(213, 222, 241, 0.25);
  border-radius: 5px;
}

.buttonWrapper {
  -webkit-box-pack: justify;
  -ms-flex-pack: justify;
  -webkit-box-align: center;
  -ms-flex-align: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-align-items: center;
  align-items: center;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

.caret {
  margin-left: 10px;
}

.dropdownContainer {
  position: relative;
}

<div
  className="container"
>
  <div>
    <button
      className="button-box"
      onClick={[Function]}
      type="button"
    >
      <span
        className="buttonWrapper"
      >
        dropdown
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-chevron-down fa-w-14 fa-sm caret"
          data-icon="chevron-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </span>
    </button>
    <div
      className="dropdownContainer"
    />
  </div>
</div>
`;

exports[`CaseDetail Component should render with initial value 1`] = `
.container {
  display: inline-block;
  font-size: 12px;
}

.button-box {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 0 7px;
  line-height: 32px;
  color: #D5DEF1;
  background-color: #FFF;
  border: 1px solid rgba(213, 222, 241, 0.25);
  border-radius: 5px;
}

.buttonWrapper {
  -webkit-box-pack: justify;
  -ms-flex-pack: justify;
  -webkit-box-align: center;
  -ms-flex-align: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-align-items: center;
  align-items: center;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

.dropdownContainer {
  position: relative;
}

.chip {
  cursor: pointer;
  padding-left: 7px;
}

.caret-caretRotate {
  margin-left: 10px;
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.dropdownBody-box {
  position: absolute;
  left: 0px;
  top: 5px;
  min-width: 100%;
  -moz-box-sizing: border-box;
  box-sizing: border-box;
  background: #FFF;
  padding: 12px;
  box-shadow: 0 1px 5px 0 rgba(213, 222, 241, 0.25);
  z-index: 300;
  border: 1px solid rgba(213, 222, 241, 0.25);
  border-radius: 5px;
}

.option {
  -webkit-box-align: center;
  -ms-flex-align: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-align-items: center;
  align-items: center;
  cursor: pointer;
  margin-bottom: 10px;
}

.optionLabel {
  display: inline-block;
  margin-left: 7px;
  text-transform: initial;
  white-space: nowrap;
}

.clear {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 10px 0 0;
  border: 0px;
  font-size: 14px;
  font-weigh: 300px;
  background: transparent;
  color: #D5DEF1;
  cursor: pointer;
}

.block-checkboxType {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  position: relative;
  outline: 0px;
  cursor: pointer;
  width: 18px;
  height: 18px;
  display: inline-block;
}

.block-checkboxType:checked:before {
  background-color: #000;
  border: 1px solid #000;
}

.block-checkboxType:checked:after {
  background-position: center;
  background-size: cover;
  background-image: url(check.svg);
}

.block-checkboxType:disabled {
  opacity: 0.45;
}

.block-checkboxType:before {
  content: "";
  position: absolute;
  display: block;
  width: 18px;
  height: 18px;
  border: 1px solid rgba(213, 222, 241, 0.7);
  border-radius: 4px;
}

.block-checkboxType:after {
  content: "";
  position: absolute;
  display: block;
  width: 18px;
  height: 18px;
  border: 1px solid transparent;
  border-radius: 4px;
}

<div
  className="container"
>
  <div>
    <button
      className="button-box"
      onClick={[Function]}
      type="button"
    >
      <span
        className="buttonWrapper"
      >
        dropdown
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-chevron-down fa-w-14 fa-sm caret-caretRotate"
          data-icon="chevron-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </span>
    </button>
    <span
      className="chip"
      data-testid="chip"
      onClick={[Function]}
    >
      one
       
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-times fa-w-11 "
        data-icon="times"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 352 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </span>
    <div
      className="dropdownContainer"
    >
      <div
        className="dropdownBody-box"
      >
        <div
          className="option"
          data-testid="dropdown-option"
          onClick={[Function]}
        >
          <input
            checked={true}
            className="block-checkboxType"
            disabled={false}
            readOnly={true}
            type="checkbox"
          />
          <span
            className="optionLabel"
          >
            1
          </span>
        </div>
        <div
          className="option"
          data-testid="dropdown-option"
          onClick={[Function]}
        >
          <input
            checked={false}
            className="block-checkboxType"
            disabled={false}
            readOnly={true}
            type="checkbox"
          />
          <span
            className="optionLabel"
          >
            2
          </span>
        </div>
        <button
          className="clear"
          onClick={[Function]}
          type="button"
        >
          Clear
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`CaseDetail Component should render without showing chips 1`] = `
.container {
  display: inline-block;
  font-size: 12px;
}

.button-box {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 0 7px;
  line-height: 32px;
  color: #D5DEF1;
  background-color: #FFF;
  border: 1px solid rgba(213, 222, 241, 0.25);
  border-radius: 5px;
}

.buttonWrapper {
  -webkit-box-pack: justify;
  -ms-flex-pack: justify;
  -webkit-box-align: center;
  -ms-flex-align: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-align-items: center;
  align-items: center;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

.caret {
  margin-left: 10px;
}

.dropdownContainer {
  position: relative;
}

<div
  className="container"
>
  <div>
    <button
      className="button-box"
      onClick={[Function]}
      type="button"
    >
      <span
        className="buttonWrapper"
      >
        dropdown (1)
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-chevron-down fa-w-14 fa-sm caret"
          data-icon="chevron-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </span>
    </button>
    <div
      className="dropdownContainer"
    />
  </div>
</div>
`;
